### PR TITLE
fix: make sure multiple functions are not using the same artifact

### PR DESCRIPTION
**Pull request by scadu**
Closes: #4128

Update: As stated in https://github.com/serverless/serverless/pull/8891#discussion_r591185530 we're going to refactor related tests with `runServerless` before merging this one.

### CODE REVIEW

File: lib/plugins/aws/package/compile/functions.js [hunk 1]
```diff
@@ -221,7 +221,7 @@ class AwsCompileFunctions {
 
       functionResource.Properties.Code.S3Key = `${
         this.serverless.service.package.artifactDirectoryName
-      }/${artifactFilePath.split(path.sep).pop()}`;
+      }/${artifactFilePath.split(path.sep).splice(-2).join(path.sep)}`;
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/functions.js:224]: This should properly differentiate for artifacts nested one level, what about if someone uses more deeply nested `artifact`s? I know it's quite exotic example, but I was wondering, is there any danger in using the whole provided artifact path ?

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:224]: I've quickly tested this example:

```
functions:
  hello:
    handler: my-function.handler
    package:
      artifact: other/nested/artifact/testservicename.zip

  hello2:
    handler: my-function.handler
    package:
      artifact: very/nested/artifact/testservicename.zip
```

And the S3Keys are identical for both functions. 

@medikoo do you think it's fine to support one-level nesting or we should support all possible nestings?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @scadu :bow: I have one question related to the approach, please let me know what do you think and what was the reasoning behind selected approach :bow:

File: lib/plugins/aws/package/compile/functions.js (hunk 1 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:224]: @pgrzesik That was the solution suggested in https://github.com/serverless/serverless/issues/4128#issue-251476320, but the implementation might be indeed flawed, or at least won't support deeper nesting levels.
I'll think about more universal approach in the meantime.

Comment (medikoo) [reply lib/plugins/aws/package/compile/functions.js:224]: I believe we should support all possible nesting.

I think right solution is to simply strip `<service-path>/.serverless` from a full path. Still we need to also confirm how it works with custom artifacts (they're not put into `.serverless`). Maybe right logic is to:
- If file is in `<service-path>/.serverless/` strip `<service-path>/.serverless/`
- Otherwise strip `<service-path>` (and throw if artifact is not even in `<service-path>`, that I believe should not be supported

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:224]: Thanks for explanation @scadu :bow: 

@medikoo for custom artifacts I believe the path is exactly the same as defined in `serverles.yml` so in the case of the example I used, path is equal to `other/nested/artifact/testservicename.zip` In such cases we can just use it as-is I believe

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:224]: @pgrzesik I tested your example on the current master and my fork and the S3Key is not identical in this case:
```
❯ Select-String "S3Key" .\.serverless\serverless-state.json

.serverless\serverless-state.json:182:                "S3Key": "serverless/zo
mgservice/dev/1612552187918-2021-02-05T19:09:47.918Z/other/nested/artifact/ar
tifact.zip"
.serverless\serverless-state.json:207:                "S3Key": "serverless/zo
mgservice/dev/1612552187918-2021-02-05T19:09:47.918Z/very/nested/artifact/art
ifact.zip"
```

`serverless.yml`:
```yml
service: zomgservice

configValidationMode: error
frameworkVersion: '*'

provider:
  name: aws

functions:
  hello:
    handler: my-function.handler
    package:
      artifact: other/nested/artifact/artifact.zip

  hello2:
    handler: my-function.handler
    package:
      artifact: very/nested/artifact/artifact.zip
```
So, at this moment I'm not even able to reproduce the issue. However, I'll continue with different scenarios to make sure there's no clash and try with @medikoo's suggestions.

I marked the PR as draft for now.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:224]: That's interesting - however if I'm not mistaken, you're on Windows, right? If that's the case, `other/nested/artifact/artifact.zip` will be treated as a single item when splitted on `path.sep`. I've just tested again and got identical keys. Let me know if that's the case on your side :bow:

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: @pgrzesik Indeed, it behaves differently on Windows and Linux. The current state of the PR should behave the same on Windows and Linux. I didn't introduce any conditionals mentioned above.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Thanks @scadu - I've tested that approach locally and while it's working perfectly for explicitly defined custom artifacts as in my example, it's breaking for functions that use artifacts built with the Framework

For this definition:
```
hello3:
    handler: my-function.handler
    package:
      individually: true
```

That's the `S3Key` that is being generated on my machine:
```
serverless/test--servicename/dev/1612786076339-2021-02-08T12:07:56.339Z//home/pgrzesik/Resources/work-serverless/reproduce/.serverless/hello3.zip
```

which is invalid. I believe the best way would be to take @medikoo suggestion to first strip everything up to `.serverless` directory in path. Please let me know what do you think

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: No worries, it's a tricky one :100:

File: lib/plugins/aws/package/compile/functions.js [hunk 2]
```diff
@@ -219,9 +218,20 @@ class AwsCompileFunctions {
...
+      // Check if it's not a custom artifact
+      if (serverlessPackagedArtifact > -1) {
+        functionResource.Properties.Code.S3Key = `${
+          this.serverless.service.package.artifactDirectoryName
+        }/${artifactFilePath
+          .split(path.sep)
+          .splice(-1, `${serverlessPackagedArtifact}`)
+          .join(path.sep)}`;
+      } else {
+        functionResource.Properties.Code.S3Key = `${this.serverless.service.package.artifactDirectoryName}/${artifactFilePath}`;
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/functions.js:232]: I think in this situation we should still account for the fact that `artifactFilePath` can start with a `/` or `\` that should be escaped and for S3 I believe we might need to split on `path.sep`, but join on `/` to be consistent, otherwise we might end up in a situation where deploying the same service from Windows and Linux machine will result in different S3 Keys. @medikoo what is your opinion here?

File: lib/plugins/aws/package/compile/functions.js [hunk 3]
```diff
@@ -219,9 +218,20 @@ class AwsCompileFunctions {
         ? this.serverless.service.package.deploymentBucket
         : { Ref: 'ServerlessDeploymentBucket' };
 
-      functionResource.Properties.Code.S3Key = `${
-        this.serverless.service.package.artifactDirectoryName
-      }/${artifactFilePath.split(path.sep).pop()}`;
+      const serverlessPackagedArtifact = artifactFilePath.split(path.sep).indexOf('.serverless');
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/functions.js:221]: I'm wondering if we could replace the logic that detects if there's `serverless` in path with something like:

```
artifactFilePath.split('.serverless').pop()
```

What do you think?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
thanks @scadu :bow: I have a few questions/comments, please see them and let me know what do you think.

File: lib/plugins/aws/package/compile/functions.js (hunk 2 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/functions.js:232]: I guess it's a good point. I think we may simply unconditionally replace `\` with `/`

File: lib/plugins/aws/package/compile/functions.js (hunk 3 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:221]: As far as I can see it breaks tests on Windows, while this one (after applying suggestion mentioned in https://github.com/serverless/serverless/pull/8891#discussion_r574462870) works correctly resulting in the same S3Key format on *nix and Windows.

File: lib/plugins/aws/package/compile/functions.js (hunk 2 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:232]: Done. `S3Key` format is now consistent on *nix and Windows.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:232]: Hey @scadu - did you push recent changes? If I'm not mistaken the code is the same as before

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:232]: @pgrzesik Just pushed, sorry for that.

File: lib/plugins/aws/package/compile/functions.js (hunk 3 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:221]: Hello :wave: I've checked the previous commits and I see that you've using split on `.serverless/` which will indeed break as it includes a `/` - I think that using `.serverless` instead of `.serverless/` when splitting should work correctly here

File: lib/plugins/aws/package/compile/functions.js (hunk 2 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:232]: No worries @scadu :100: I've just checked the change and I believe it's still going to be problematic when the `artifactFilePath` is an absolute path, as it will create an invalid S3Key e.g. if `artifactFilePath` would be `/path/to/file`, in this case it would result in an `S3Key` with two slashes next to each other. I think we should try to transform absolute paths by stripping the `/` at the beginning on the equivalent on Windows. What do you think?

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @scadu - I believe we still have some problematic edge-cases, please see my comments and let me know what do you think.

File: lib/plugins/aws/package/compile/functions.js (hunk 2 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:232]: I'll take a look.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:232]: Thank you :bow:

File: lib/plugins/aws/package/compile/functions.js (hunk 3 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: I'm afraid it'll produce S3Key like `some/path//path/to/artifact/artifact.zip` or `some/path/\path/to/artifact/artifact.zip` (depending on the OS). 

```js
artifactFilePath.split(`.serverless${path.sep}`).pop()
```

is what I used for now.

File: lib/plugins/aws/package/compile/functions.js (hunk 2 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: There's still hackery but I couldn't figure out anything more elegant. Trying to get tests working again.
The `path.root` is removed now. Not sure if we would also like to convert Windows path like `some\\path\\to\\artifact.zip` to `some/path/to/artifact.zip` for S3Key.

File: lib/plugins/aws/package/compile/functions.js [hunk 4]
```diff
@@ -219,9 +218,16 @@ class AwsCompileFunctions {
         ? this.serverless.service.package.deploymentBucket
         : { Ref: 'ServerlessDeploymentBucket' };
 
-      functionResource.Properties.Code.S3Key = `${
-        this.serverless.service.package.artifactDirectoryName
-      }/${artifactFilePath.split(path.sep).pop()}`;
+      if (artifactFilePath.includes(`.serverless${path.sep}`)) {
```
Comment (pgrzesik) [inline lib/plugins/aws/package/compile/functions.js:221]: I think we could avoid that if statement here. What if we would do something like this (variable names are probably not the best below, I just want to illustrate the approach):

```
const pathWithoutServerless = artifactFilePath.split('.serverless').pop()
```

The above will ensure that for provided paths, it will be assigned to `pathWithoutServerless` and for situations where the artifact from `.serverless` is used, we'll strip it and only use the part of path inside `.serverless` dir. I think we don't need to include `path.sep` in the split because we can uniformly handle that later.

```
const pathWithUnixSlashes = pathWithoutServerless.replace(/\\/g, '/');
const pathWithoutLeadingSlash = pathWithUnixSlashes.startswith('/') ? pathWithUnixSlashes.substring(1) : pathWithUnixSlashes;
```

Here we will have the path with replaced slashes + removed leading slash if it exists. However after second though I guess it might break for paths on Windows like `C:\\test\\test` - does the approach with replacing `path.parse(artifactFilePath`.root)` resolve replacing the `C:\\` ?

What do you think about the above approach? I think it's a bit simplified vs the one that uses `if` branching.

File: lib/plugins/aws/package/compile/functions.js (hunk 2 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Hello :wave: Thank you @scadu I've posted a small "proposal" and I'd love to hear your opinion on the approach. As for convering paths - I think we should ensure it's always converted to using forward slashes for `S3Key`. :+1:

File: lib/plugins/aws/package/compile/functions.js (hunk 4 above)
Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js:221]: It'd break S3Keys on Windows for sure. That's why I started tinkering with `path.parse` to have a unified method for removing a path root and only this, since I don't see any other way to meet our goal and ensure unique S3Key for custom artifacts, especially with absolute path provided.

I was trying to save a path in the Unix standard to prevent playing with regular expressions, without luck so far.
I guess we might try the suggested solution and ensure a path root is removed (via `path.parse`).
I'll mess around and we'll see how it went.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js:221]: Thank you @scadu :bow:

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: I just pushed the latest changes and it seems to be working as expected (tested on Linux and Windows). 
Now I'm just trying to get tests working again, without creating much boilerplate. 
I'll keep the PR as a draft until I figure it out if that's okay with you.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Sure thing - thank you for all your work on this one  :bow:

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: Small update: I still have a few tests failing (three - to be precise). I'll keep working on fixing those.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Thanks for an update @scadu :bow:

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: It seems that I've hit a dead-end again. I've managed to get most of the tests working without creating much boilerplate, however, I'm still struggling with three of them:

```
  1) AwsCompileFunctions
       #downloadPackageArtifacts()
         should download the file and replace the artifact path for function packages:
      AssertionError: expected 'Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/9350ee01b157468c/new-service.zip' to equal 's3-hosted-artifact.zip'
      + expected - actual

      -Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/9350ee01b157468c/new-service.zip
      +s3-hosted-artifact.zip

      at C:\Users\ljend\Code\serverless\test\unit\lib\plugins\aws\package\compile\functions.test.js:123:37
 ```
      
 ```
   2) AwsCompileFunctions
       #compileFunctions()
         should use function artifact if individually:

      AssertionError: expected 'somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/test.zip' to deeply equal 'somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/new-service.zip'
      + expected - actual

      -somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/test.zip
      +somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/new-service.zip

      at C:\Users\ljend\Code\serverless\test\unit\lib\plugins\aws\package\compile\functions.test.js:170:64
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

```
  3) AwsCompileFunctions
       #compileFunctions()
         should honour provider.iam.role option when set:
     ServerlessError: Serverless command "C:\Users\ljend\Code\serverless\test\unit\lib\plugins\aws\package\compile\functions.test.js" not found. Did you mean "credentials"? Run "serverless help" for a list of all available commands.
      at commandOrAlias.reduce.commands (lib\classes\PluginManager.js:459:15)
      at Array.reduce (<anonymous>)
      at PluginManager.getCommand (lib\classes\PluginManager.js:422:27)
      at PluginManager.validateCommand (lib\classes\PluginManager.js:549:10)
      at Serverless.run (lib\Serverless.js:234:24)
      at C:\Users\ljend\Code\serverless\node_modules\@serverless\test\run-serverless.js:261:36
      at C:\Users\ljend\Code\serverless\node_modules\@serverless\test\setup-run-serverless-fixtures-engine.js:33:22
      at Context.<anonymous> (test\unit\lib\plugins\aws\package\compile\functions.test.js:245:30)
```

For the rest of the tests, [this conditional works perfectly](https://github.com/serverless/serverless/pull/8891/files#diff-b0a2f2c747c874b3ac996b289739e323313d074cf998a8cf83910d2d9380b181R76-R92).
The first test failing has specified `awsCompileFunctions.serverless.service.functions[functionName].package.artifact` which is some hint (though I'm not sure how to pass it properly), but I have no clue about the second one (`should use function artifact if individually`), not to mention the last one.
It looks like it's a small thing that I can't spot.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Sorry for delayed response - where do you see these failing tests? I've just checked the last pipeline and it seems like only one of them is failing?

As for the adjustments - if we're going this far with changing the tests, I'm afraid that we should consider replacing them with `runServerless`-based ones which will heavily impact the scope of this PR - from what I see the current ones are quite fragile and I'm not sure about their reliability to be honest. What do you think?

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: One of them is failing [for unit test ran on Node 14](https://github.com/serverless/serverless/pull/8891/checks?check_run_id=2024336592), another two can be seen in the step of [isolated unit test ran on Node 12](https://github.com/serverless/serverless/pull/8891/checks?check_run_id=2024336388).

Regarding reliability, do you mean the conditionals introduced in this PR or the fact that `runServerless` is not used everywhere in `functions.test.js`? Nevertheless, I'll try to include #8523 into this PR.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Thanks, I've missed the fact that they're inconsistently failing across different Node versions. 

As for reliability - I'm talking more about the genenral approach used in existing tests - they heavily depend on manually constructing fragile examples for tests which makes it then hard to refactor. Ideally would be to first refactor the package-related tests in `functions.test.js` (not necesarilly all of them because it's a lot of work) to use `runServerless` and then introduce that change - we made a decision to not introduce extra tests with old approach and I feel like with the conditionals and changes here we're kinda adding to "old" tech debt of our testing suite. Please let me know if you're up for that work, as I know it grows the initial scope of the PR and I totally understand if you're not interested.

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: Yeah, that totally makes sense. I just thought that I'd finish this one and pick #8523 but now it seems that the initial @medikoo's suggestion was more reasonable (back then we also had some edge cases to cover).

If it's not urgent (and assuming that the issue is opened for over three years now it isn't but I don't have the same context), I'd take a closer look into rewriting related tests in `functions.test.js` to use `runServerless`, especially when we got this far already.
@pgrzesik let me know what's your position on this issue.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: That would be perfect @scadu as I think to wrap this one up, we will need updated test suite and I believe it makes a lot of sense to refactor it as a part of separate PR. This one definitely can wait :+1: Thank you for your efforts and help @scadu :bow:

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Sorry for not being clear - My thinking was that we could put this PR on a pause for a moment, cover necessary parts from #8523 as a part of a separate PR, merge that PR and then come back to this one, relying on new, `runServerless`-based tests only. I know it prolongs this PR, but if you're okay with it, I believe that would be 'cleanest' approach. Let me know that do you think :raised_hands: Thanks for the kind words as well :rocket:

Comment (scadu) [reply lib/plugins/aws/package/compile/functions.js]: @pgrzesik oh, right. I'll update #8523 with this information to make prevent duplicated work.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/functions.js]: Perfect, thank you :clap:

**Comment by medikoo** [comment]
@scadu what exactly is state of that? We'll be very happy to review once it's finalized

**Comment by scadu** [comment]
@medikoo all edge cases we've found are covered. I got stuck with few tests however and then we've agreed with @pgrzesik (see https://github.com/serverless/serverless/pull/8891#discussion_r591185530) to refactor packaging-related ones as a part of #8523. This was two weeks ago and haven't had a chance to move it forward.

**Comment by scadu** [comment]
@pgrzesik @medikoo I tried to get into #8523 with not much luck. Looks like the pandemic got me down a little, so I'd like to back off from working on #4128 and related #8523. Sorry for ping-ponging like this. I guess this PR can be closed.

**Comment by pgrzesik** [comment]
Hello @scadu - that's totally understandable - thanks for your efforts so far :bow: Hopefully you'll have a speedy recovery :crossed_fingers:
